File: sdk/runanywhere-kotlin/src/commonMain/kotlin/com/runanywhere/sdk/events/EventPublisher.kt
package com.runanywhere.sdk.events

import com.runanywhere.sdk.foundation.SDKLogger
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

/**
 * Simple event router for the SDK.
 * Mirrors iOS EventPublisher.
 *
 * Just call `track(event)` - the router decides where to send it
 * based on the event's `destination` property.
 *
 * Usage:
 * ```kotlin
 * EventPublisher.track(LLMEvent.GenerationCompleted(...))
 * ```
 */
object EventPublisher {

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
    private val logger = SDKLogger("EventPublisher")

    /**
     * Analytics queue for telemetry events.
     * Set during SDK initialization via [initialize].
     */
    private var analyticsEnqueuer: AnalyticsEnqueuer? = null

    /**
     * Interface for analytics queue integration.
     * Implement this to receive events destined for analytics.
     */
    fun interface AnalyticsEnqueuer {
        suspend fun enqueue(event: SDKEvent)
    }

    /**
     * Initialize with analytics queue (call during SDK startup)
     */
    fun initialize(analyticsQueue: AnalyticsEnqueuer?) {
        analyticsEnqueuer = analyticsQueue
        logger.info("Initialized with analytics queue: ${analyticsQueue != null}")
    }

    /**
     * Track an event. Routes automatically based on event.destination.
     *
     * @param event The event to track
     */
    fun track(event: SDKEvent) {
        val destination = event.destination

        // Route to EventBus (public) if not analytics-only
        if (destination != EventDestination.ANALYTICS_ONLY) {
            EventBus.publishSDKEvent(event)
        }

        // Route to Analytics (telemetry) if not public-only
        if (destination != EventDestination.PUBLIC_ONLY) {
            analyticsEnqueuer?.let { enqueuer ->
                scope.launch {
                    try {
                        enqueuer.enqueue(event)
                    } catch (e: Exception) {
                        logger.error("Failed to enqueue event for analytics: ${e.message}")
                    }
                }
            }
        }
    }

    /**
     * Track an event asynchronously (for use in suspend contexts).
     *
     * @param event The event to track
     */
    suspend fun trackAsync(event: SDKEvent) {
        val destination = event.destination

        // Route to EventBus (public) if not analytics-only
        if (destination != EventDestination.ANALYTICS_ONLY) {
            EventBus.publishSDKEvent(event)
        }

        // Route to Analytics (telemetry) if not public-only
        if (destination != EventDestination.PUBLIC_ONLY) {
            analyticsEnqueuer?.let { enqueuer ->
                try {
                    enqueuer.enqueue(event)
                } catch (e: Exception) {
                    logger.error("Failed to enqueue event for analytics: ${e.message}")
                }
            }
        }
    }

    /**
     * Check if analytics is configured.
     */
    val isAnalyticsConfigured: Boolean
        get() = analyticsEnqueuer != null
}

/**
 * Convenience extension to track events directly.
 */
fun SDKEvent.track() = EventPublisher.track(this)

/**
 * Convenience extension to track events asynchronously.
 */
suspend fun SDKEvent.trackAsync() = EventPublisher.trackAsync(this)

File: sdk/runanywhere-flutter/packages/runanywhere/lib/infrastructure/events/event_publisher.dart
/// Event Publisher
///
/// Simple event router. Call track(event) and it handles the rest.
/// Corresponds to iOS SDK's EventPublisher.swift
library event_publisher;

import 'dart:async';

import 'package:runanywhere/public/events/event_bus.dart';
import 'package:runanywhere/public/events/sdk_event.dart';

/// Event destination options
enum EventDestination {
  /// Send to public EventBus only
  publicOnly,

  /// Send to analytics/telemetry only
  analyticsOnly,

  /// Send to both public EventBus and analytics
  both,
}

/// Simple event router for the SDK.
///
/// Mirrors iOS `EventPublisher` from RunAnywhere SDK.
/// Just call `track(event)` - the router decides where to send it.
///
/// Usage:
/// ```dart
/// EventPublisher.shared.track(LLMEvent.generationCompleted(...));
/// ```
class EventPublisher {
  // MARK: - Singleton

  /// Shared instance
  static final EventPublisher shared = EventPublisher._();

  EventPublisher._({EventBus? eventBus})
      : _eventBus = eventBus ?? EventBus.shared;

  // MARK: - Dependencies

  final EventBus _eventBus;

  // MARK: - Track

  /// Track an event. Routes automatically based on event.destination.
  void track(SDKEvent event) {
    final destination = event.destination;

    // Route to EventBus (public)
    if (destination != EventDestination.analyticsOnly) {
      _eventBus.publish(event);
    }

    // Route to Analytics (telemetry) - handled via FFI/DartBridge
    if (destination != EventDestination.publicOnly) {
      // TODO: Route to DartBridge.Telemetry
    }
  }

  /// Track an event asynchronously (for use in async contexts).
  Future<void> trackAsync(SDKEvent event) async {
    track(event);
  }
}
